### PR TITLE
Run tests in GitHub Actions

### PR DESCRIPTION
**Pull request by paulproteus**
Additionally, add some cleanup within the existing test suite to
handle non-determinism of which tests run in which order.

### CODE REVIEW

**Comment by paulproteus** [comment]
@freakboy3742 I request your review, and if you like it, please merge!

I intend to continue to pick pieces off of #5 until they're all merged in fairly minimal, self-contained PRs. Every time I make one of these fairly minimal, self-contained PRs, I end up learning new things about the codebase. This time, I learned that if you run the test suite in GitHub Actions in a different order than the order it runs on my laptop, you end up with tests that fail for seemingly mysterious reasons. Hence the new context manager.

Happy to address feedback from your review.

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+on: [push]
```
Comment (freakboy3742) [inline .github/workflows/test.yaml]: It turns out `push` only triggers when you *open* a PR (or when you merge), so `pull_request` is more useful here.

**Comment by paulproteus** [comment]
Hi! Somehow I have a lot of friendly, excited comments on your comments. Feel free to read them, or ignore them.

> Looks good. I'm sure we can add more as time goes by (like adding beefore checks for linting).

Big +1 from me. :)

> I'm interested why I've never seen the test ordering problem before. Makes sense why it might be a problem; it's just interesting to me that I've never seen it.

It's weird! I ran into in GitHub Actions, but I didn't run into it on my laptop. This was a reliable experience for me. I'm going to leave it a mystery for now; I agree it's interesting that neither you nor me experience it on our machines, but GitHub Actions does.

> It turns out that I (in my capacity as repo admin) need to poke the repository to make the actions actually start triggering, so I've made a couple of small changes. 

Fascinating! I didn't run into any such thing on my fork. Would be curious to know what you ran into, or feel free to not answer this if you think this question of mine is a boring distraction. Is this about allowing GH Actions to run for people without `beeware/rubicon-java` repo permissions?

> One (the trigger for the action) I've flagged inline as a coment

I don't see the comment -- can you point me to it? What I do see is two build YAML files which are near-duplicates of each other.

I have just discovered that you can have one GitHub Actions workflow apply in multiple situations, per https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#example-using-multiple-events-with-activity-types-or-configuration , so I am likely to submit a follow-up implementing that.

> I've also added a placeholder for matrix builds (which we're going to need very soon)

I was avoiding matrix builds in this PR as part of a personal focus on YAGNI/minimalism. Having matrix builds in here is OK with me, too. :)

> a separate build_status workflow that will re-run the tests when a PR is merged.

Cool!

**Comment by paulproteus** [comment]
(And thanks for the review & merge!)

**Comment by freakboy3742** [comment]
The comment I referenced is [https://github.com/beeware/rubicon-java/pull/9/files/59ead3e416b0a85ec00a762c38adbf62aa58dec6#r356425504](here). 

As for the needing to poke it myself... that's interesting. Looking at the [action history](https://github.com/beeware/rubicon-java/actions), I can see that there were a bunch of successful builds triggered by your commits... but none of those build seem to correspond to the commit hashes that are actually in this PR - it looks like the were all on an "github-actions-on-mac" build. The two commits attributed to you on *this* PR didn't trigger builds. I assumed that was because I was the owner... but maybe not? Anyway... it's a moot point now.

As for the multiple workflow files - that's actually with a purpose. The pattern I've used on other repos (cf [toga](https://github.com/beeware/toga/tree/master/.github/workflows) is that the CI workflow is a full test on all platforms, preceded by code quality checks (linting etc); the Build status workflow is a smoke test on a single representative platform that can be used to flag whether builds are succeeding or failing. 

The primary reason for this is the build status badge in the README. You can't have a build status badge on your README that is driven by the CI workflow, because *any* PR that fails will alter the state of the badge, regardless of whether it's merged. You need a distinct workflow that only triggers on commits to master. If CI passes on a PR, we should have a reasonable degree of confidence that result will persist into master; but we need to trigger a distinct "yeah, really, everything is OK" build on master to provide data to drive the badge. 

As for YAGNI - I can understand that as a motivation, but this is a case where we're very definitely GNI... :-)

If you can find a better way to manage this, I'm open to suggestion - but in terms of billable time, I don't consider this a high priority for the moment.

**Pull request merged**